components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/TokenResponseTypeHandler.java

```diff
@@ -135,170 +135,179 @@ public OAuth2AuthorizeRespDTO issue(OAuthAuthzReqMessageContext oauthAuthzMsgCtx
-
-                long refreshTokenExpiryTime = OAuth2Util.getRefreshTokenExpireTimeMillis(existingAccessTokenDO);
+            // if access token is not found in cache, check if the last issued access token is still active and valid
+            // in the database
+            if (existingAccessTokenDO == null) {
 
-                if (OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(
-                        existingAccessTokenDO.getTokenState()) && (expiryTime > 0 || expiryTime < 0)) {
+                existingAccessTokenDO = OAuthTokenPersistenceFactory.getInstance().getAccessTokenDAO()
+                        .getLatestAccessToken(consumerKey, authorizationReqDTO.getUser(), userStoreDomain, scope,
+                                false);
+                if (existingAccessTokenDO != null) {
 
-                    // token is active and valid
                     if (log.isDebugEnabled()) {
-                        if (expiryTime > 0) {
-                            log.debug("Access token is valid for another " + expiryTime + "ms");
-                        } else {
-                            log.debug("Infinite lifetime Access Token found in cache");
-                        }
-                    }
-
-                    if (cacheEnabled) {
-                        OAuthCache.getInstance().addToCache(cacheKey, existingAccessTokenDO);
-                        if (log.isDebugEnabled()) {
-                            log.debug("Access Token was added to cache for cache key : "
-                                    + cacheKey.getCacheKeyString());
-                        }
-                    }
-
-                    respDTO.setAccessToken(existingAccessTokenDO.getAccessToken());
-
-                    if (expiryTime > 0) {
-                        respDTO.setValidityPeriod(expiryTime / 1000);
-                    } else {
-                        respDTO.setValidityPeriod(Long.MAX_VALUE / 1000);
+                        log.debug("Retrieved latest Access Token for Client ID : " + consumerKey + ", User ID :"
+                                + authorizedUser + " and Scope : " + scope + " from database");
                     }
 
-                    respDTO.setScope(oauthAuthzMsgCtx.getApprovedScope());
-                    respDTO.setTokenType(existingAccessTokenDO.getTokenType());
+                    long expiryTime = OAuth2Util.getTokenExpireTimeMillis(existingAccessTokenDO);
+                    long refreshTokenExpiryTime = OAuth2Util.getRefreshTokenExpireTimeMillis(existingAccessTokenDO);
 
-                    // we only need to deal with id_token and user attributes if the request is OIDC
-                    if (isOIDCRequest(oauthAuthzMsgCtx)) {
-                        buildIdToken(oauthAuthzMsgCtx, respDTO);
-                    }
+                    if (OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(existingAccessTokenDO.getTokenState())
+                            && (expiryTime > 0 || expiryTime < 0)) {
```

Comment:
expiryTime != 0 ?

Reply:
This class is deprecated and I did minimal changes to get the respective issue fixed rather than existing logic.
This logic was improved in the ResposeHandlerUtil class which will be used going forward.